hydra/core/plugins.py

```diff
@@ -184,20 +184,16 @@ def _scan_all_plugins(
                     import_time = timer()
 
                     with warnings.catch_warnings(record=True) as recorded_warnings:
-                        if sys.version_info < (3, 10):
-                            m = importer.find_module(modname)  # type: ignore
-                            assert m is not None
-                            loaded_mod = m.load_module(modname)
+                        spec = importer.find_spec(modname, None)
+                        assert spec is not None
+                        if modname in sys.modules:
+                            loaded_mod = sys.modules[modname]
                         else:
-                            spec = importer.find_spec(modname)
-                            assert spec is not None
-                            if modname in sys.modules:
-                                loaded_mod = sys.modules[modname]
-                            else:
-                                loaded_mod = importlib.util.module_from_spec(spec)
-                            if loaded_mod is not None:
-                                spec.loader.exec_module(loaded_mod)
-                                sys.modules[modname] = loaded_mod
+                            loaded_mod = importlib.util.module_from_spec(spec)
+                        if loaded_mod is not None:
+                            sys.modules[modname] = loaded_mod
+                            assert spec.loader
+                            spec.loader.exec_module(loaded_mod)
```

Comment:
Shouldn't the `exec_module` line come before the `sys.modules[modname] = loaded_mod` line?
I'm worried that the current line order might leave `sys.modules` in an inconsistent state, with an improperly initialized module, if `exec_module` raises an exception.

Reply:
It has to be that way around so that if the plugin has any kind of circular imports, the module is findable by its submodules. See the bottom of [this example](https://docs.python.org/3/library/importlib.html#approximating-importlib-import-module).

It does sound sensible for failed imports to be removed. The docs don't mention anything about removing them but a quick experimentation shows that `import` and `importlib.import_module()` adds the module at the start and then removes it again on failure.
```python
# broken_plugin.py
import sys
print(sys.modules["broken_plugin"])
assert 0
```
```python
>>> import broken_plugin
<module 'broken_plugin' from '/tmp/broken_plugin.py'>
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/broken_plugin.py", line 3, in <module>
    assert 0
AssertionError

>>> import sys
>>> "broken_plugin" in sys.modules
False
```
Same story if I swap `import` for `importlib.import_module()`.

I can add a `sys.modules.pop(modname, None)` to the `except` block if you prefer?

Reply:
Hmm, I didn't think about submodules.
I think cleaning up broken submodules could be difficult. It's easy to do `pop(modname, None)`, but `pop(submodule_name, None)` is more challenging. Maybe it's better keep the code simple and let the user worry about cleaning up the broken plugins...